sources/api/prairiedog/src/main.rs

```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+    trace!("Boot settings: {:?}", boot_settings);
+    let bootconfig = to_string(&boot_settings).context(error::SerializeBootConfigSnafu)?;
+    trace!("Serializied boot config string: {}", bootconfig);
+    let initrd = generate_initrd(&bootconfig.into_bytes())?;
+    trace!("Writing initrd image file: {:?}", initrd);
+    tokio::fs::write(BOOTCONFIG_INITRD_PATH, &initrd)
+        .await
+        .context(error::WriteInitrdSnafu)?;
+    Ok(())
+}
+
+async fn get_boot_config_settings<P>(socket_path: P) -> Result<BootSettings>
+where
+    P: AsRef<Path>,
+{
+    let boot_settings = schnauzer::get_settings(socket_path)
+        .await
+        .context(error::RetrieveSettingsSnafu)?
+        .settings
+        .and_then(|s| s.boot)
+        .context(error::MissingSettingsSnafu)?;
+    Ok(boot_settings)
+}
+
+fn generate_initrd(bootconfig: &[u8]) -> Result<Vec<u8>> {
+    let mut bootconfig_size = bootconfig.len();
+    // Calculate checksum of the boot config file
+    let mut checksum: u32 = 0;
+    for byte in bootconfig {
+        checksum = checksum
+            .checked_add(u32::from(*byte))
+            .context(error::AddU32OverflowSnafu)?;
+    }
+    // Bottlerocket does not use an initrd, so the base initrd image size is 0
+    let total_size = BASE_INITRD_SIZE
+        + bootconfig_size
+        + size_of::<u32>() * 2
+        + BOOTCONFIG_MAGIC.as_bytes().len();
+    let mut initrd = bootconfig.to_owned();
+    // initrd image file needs to be 4-byte aligned, so we add padding as necessary.
+    let padding_size =
+        ((total_size + BOOTCONFIG_ALIGN - 1) & (!BOOTCONFIG_ALIGN_MASK)) - total_size;
+    trace!("Boot config size: {}", bootconfig_size);
+    trace!("Initrd total size: {}", total_size);
+    trace!("Padding size: {}", padding_size);
+    initrd.put_bytes(b'\0', padding_size);
+    bootconfig_size += padding_size;
+    // Append boot config file size (file + padding bytes)
+    initrd.put_u32_le(bootconfig_size.try_into().context(error::UsizeToU32Snafu)?);
+    // Append boot config file checksum
+    initrd.put_u32_le(checksum);
+    // Append boot config magic value
+    initrd.put(BOOTCONFIG_MAGIC.as_bytes());
+    Ok(initrd)
+}
+
+#[cfg(test)]
+mod initrd_tests {
+    use crate::generate_initrd;
+
+    static GOOD_BOOTCONFIG: &str = include_str!("../tests/data/ser_good_bootconfig");
+    static INITRD_FROM_GOOD_BOOTCONFIG: &[u8] =
+        include_bytes!("../tests/data/initrd_from_good_bootconfig");
+
+    #[test]
+    fn test_initrd_gen() {
+        let bootconfig = GOOD_BOOTCONFIG;
+        assert_eq!(
+            INITRD_FROM_GOOD_BOOTCONFIG,
+            generate_initrd(&bootconfig.as_bytes()).unwrap()
+        );
+    }
+}
+
+/// Reads `/proc/bootconfig` and populates the Bottlerocket boot settings based on the existing boot config data
+async fn generate_boot_settings() -> Result<()> {
+    let proc_bootconfig = match fs::read_to_string(PROC_BOOTCONFIG) {
+        Ok(s) => s,
+        Err(e) => {
```

Comment:
Is it possible to use pattern matching in errors, like this?

```suggestion
        Err(io::ErrorKind::NotFound(e)) => Ok(()),
        Err(e) => { fail }
```

Reply:
Unfortunately no. The matched type is of an Result which branch out to either `Ok` or `Err` but not the type of error since that's specific to the error type within the Error enum variant.